File: scan.pl
#!/usr/bin/env perl

use warnings;
use strict;

use v5.20;

use IO::Socket::INET;

use experimental qw( switch );

sub scan {
  my ($address, $port, $timeout) = @_;

  my @ips = inet_aton($address) || return "Can't resolve $address";
  my $ip = inet_ntoa(@ips) || return "Can't extract from @ips";
  my $socket = IO::Socket::INET->new(
    PeerHost => $ip,
    PeerPort => $port,
    Timeout  => $timeout,
  );

  if ($@) {
    my $err = '';
    given ($@) {
      when(/connect: timeout$/) { $err = 'closed' }
      when(/connect: connection refused$/i) { $err = 'filtered' }
      default { $err = $@  }
    }
    return $err;
  }
  $socket->close();

  return 'open';
}

say scan('google.com', 443, 3);

File: scan.php
#!/usr/bin/env php
<?php

function scan($address, $port, $timeout) {
  $ips = gethostbynamel($address);
  if ($ips === FALSE) {
    return "Unable to resolve $address";
  }

  $socket = @socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
  if ($socket === FALSE) {
    return socket_strerror(socket_last_error());
  }

  $timeout = array('sec'=>$timeout,'usec'=>$timeout * 1000);
  socket_set_option($socket,SOL_SOCKET,SO_SNDTIMEO,$timeout);
  $result = @socket_connect($socket, $ips[0], $port);
  if ($result === FALSE) {
    $err = socket_strerror(socket_last_error());
    switch ($err) {
    case 'Connection refused':
      return 'filtered';
      break;
    case 'Operation now in progress':
      return 'closed';
      break;
    default :
      print("$err\n");
      break;
    }
  }
  socket_close($socket);
  return 'open';
}

echo scan('google.com', 443, 3);
echo "\n";
?>
